server/container/util/src/main/java/reactor/retry/RetryWithAsyncCallback.java

```diff
@@ -0,0 +1,194 @@
+ *
+ *       https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package reactor.retry;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+import org.reactivestreams.Publisher;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Scheduler;
+import reactor.util.Logger;
+import reactor.util.Loggers;
+
+/**
+ * This class is a copy of reactor.retry.DefaultRetry.
+ * Its goal is to provide a way to execute an async action before retrying.
+ * To do so it provides a retryWithMono method which is the async equivalent of the synchronous method doOnRetry.
+ *
+ * This is a temporary solution as this new requirement has been exposed in an issue in the reactor project.
+ * see : https://github.com/reactor/reactor-addons/issues/220
+ *
+ */
+public class RetryWithAsyncCallback<T> extends AbstractRetry<T, Throwable> implements Retry<T> {
+
+    static final Logger log = Loggers.getLogger(RetryWithAsyncCallback.class);
+    static final Consumer<? super RetryContext<?>> NOOP_ON_RETRY = r -> {
+    };
+    static final Function<? super RetryContext<?>, Mono<?>> NOOP_ON_RETRY_MONO = r -> Mono.empty();
+
+    public static <T> RetryWithAsyncCallback<T> onlyIf(Predicate<? super RetryContext<T>> predicate) {
+        return RetryWithAsyncCallback.create(predicate).retryMax(Long.MAX_VALUE);
+    }
+
+    final Predicate<? super RetryContext<T>> retryPredicate;
+    final Consumer<? super RetryContext<T>> onRetry;
+    final Function<? super RetryContext<T>, Mono<?>> onRetryMono;
+
+    RetryWithAsyncCallback(Predicate<? super RetryContext<T>> retryPredicate,
+                           long maxIterations,
+                           Duration timeout,
+                           Backoff backoff,
+                           Jitter jitter,
+                           Scheduler backoffScheduler,
+                           final Consumer<? super RetryContext<T>> onRetry,
+                           Function<? super RetryContext<T>, Mono<?>> onRetryMono,
+                           T applicationContext) {
+        super(maxIterations, timeout, backoff, jitter, backoffScheduler, applicationContext);
+        this.retryPredicate = retryPredicate;
+        this.onRetry = onRetry;
+        this.onRetryMono = onRetryMono;
+    }
+
+    public static <T> RetryWithAsyncCallback<T> create(Predicate<? super RetryContext<T>> retryPredicate) {
+        return new RetryWithAsyncCallback<T>(retryPredicate,
+            Long.MAX_VALUE,
+            null,
```

Comment:
`Optional` would be more explicit

Reply:
It's copy/paste of code, I'd like to keep the code as close as possible of the original

Reply:
where does it came from?